fvm/evm/offchain/blocks/blocks.go

```diff
@@ -0,0 +1,122 @@
+package blocks
+
+import (
+	"fmt"
+
+	gethCommon "github.com/onflow/go-ethereum/common"
+
+	"github.com/onflow/flow-go/fvm/evm/handler"
+	"github.com/onflow/flow-go/fvm/evm/types"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+const BlockStoreLatestBlockMetaKey = "LatestBlockMeta"
+
+// Blocks facilitates access to the recent block hash values
+// and also the latest executed block meta data
+type Blocks struct {
+	chainID     flow.ChainID
+	storage     types.BackendStorage
+	rootAddress flow.Address
+	bhl         *handler.BlockHashList
+}
+
+// NewBlocks constructs a new blocks type
+func NewBlocks(
+	chainID flow.ChainID,
+	rootAddress flow.Address,
+	storage types.BackendStorage,
+) (*Blocks, error) {
+	var err error
+	blocks := &Blocks{
+		chainID:     chainID,
+		storage:     storage,
+		rootAddress: rootAddress,
+	}
+	blocks.bhl, err = handler.NewBlockHashList(
+		storage,
+		rootAddress,
+		handler.BlockHashListCapacity,
+	)
+	if err != nil {
+		return nil, err
+	}
+	// if empty insert genesis block hash
+	if blocks.bhl.IsEmpty() {
+		genesis := types.GenesisBlock(chainID)
+		err = blocks.PushBlockMeta(
+			NewMeta(
+				genesis.Height,
+				genesis.Timestamp,
+				genesis.PrevRandao,
+			))
+		if err != nil {
+			return nil, err
+		}
+		// push block hash
+		err = blocks.PushBlockHash(
+			genesis.Height,
+			types.GenesisBlockHash(chainID))
+		if err != nil {
+			return nil, err
+		}
+	}
+	return blocks, nil
+}
+
+// PushBlock pushes a new block into the storage
+func (b *Blocks) PushBlockMeta(
```

Comment:
Why meta and block hash are pushed separately? 

Do we need to avoid cases where meta is pushed but block hash is not?

Reply:
Its intentional, since the meta has to be accessible during the block execution but the hash has to be pushed at the end of the block execution. This behaviour matches the behaviour of ENs given they won't know the current block hash as more transactions are coming in.